### ISSUE TITLE
Handling of transit in valhalla_build_tiles

### ISSUE THREAD

**Issue by mesozoic-drones** [valhalla/valhalla#3557]
Hello! I saw the new proposal #3539 for GTFS support. This is a great idea! If necessary, I can update the cpp14 branch of `just_gtfs` library so that it can be conveniently integrated into Valhalla.

My question is about the support of transit in `valhalla_build_tiles`. 

In my old issue #2860 I asked about the transit pipeline and [pranavpandey1998official](https://github.com/pranavpandey1998official) built for me some transit tiles:
https://drive.google.com/drive/folders/1sRlHLRdlW3sV7jldgfTSZpO4i86mn9Zh?usp=sharing

I converted these tiles with `valhalla_convert_transit` and set `transit_dir` field in valhalla config to transit tiles path.

Then I started `valhalla_build_tiles`:
`./valhalla_build_tiles -c valhalla_conf.json san_francisco.osm.pbf`

It crashed with exception on this line:
https://github.com/valhalla/valhalla/blob/master/src/mjolnir/hierarchybuilder.cc#L347

`terminating with uncaught exception of type std::logic_error: current_level was never set`

I investigate how to make transit work, so here are my questions:

**1) What is the best way to fix this crash?**

**2) Before the crash there were errors in log:**
`No valid level for this node!`

https://github.com/valhalla/valhalla/blob/master/src/mjolnir/hierarchybuilder.cc#L472

Should I ignore or fix them? I saw the comment in code and I am not confident with the necessity of fix:
https://github.com/valhalla/valhalla/blob/master/src/mjolnir/hierarchybuilder.cc#L438-L439

I extended the log message with additional info. Example:
```
No valid level for node. nodeinfo->type() = multi_use_transit_platform. nodeinfo->is_transit() = true
No valid level for node. nodeinfo->type() = transit_egress. Inodeinfo->is_transit() = false
 No valid level for node. nodeinfo->type() = transit_station. nodeinfo->is_transit() = false
```

**3) There is a method `bool is_transit()` in `NodeInfo` class:**
https://github.com/valhalla/valhalla/blob/master/valhalla/baldr/nodeinfo.h#L212-L214

It compares node type only with the single value: `NodeType::kMultiUseTransitPlatform`. But there are also `kTransitEgress` and `kTransitStation` values in `NodeType` enum. Is it ok?

Thank you in advance!

**Comment by kevinkreiser** [comment]
@mesozoic-drones these are all great questions! i honestly didnt follow @pranavpandey1998official 's work too closely so im not quite sure what the issue is there. from your debugging i wonder why we dont have a line like this: https://github.com/valhalla/valhalla/blob/master/src/mjolnir/hierarchybuilder.cc#L441-L443 but for transit. it seems suspicious to me that we had to add a line in there for BSS connections but we should "skip" the transit connect ones. skipping them to me should mean they stay at level 2 but maybe we need to be explicit. i would at least try that just to see what happens. im actually not sure why the node type is multi use transit platform. that is not intuitive but also i didnt do the naming there. it does support everything from a bus stop on the road to a huge station with internal station geometries connecting different platforms so maybe "multi use" means "generic transit thing".

the way i remember us implementing transit connections was as follows:

```
x =========== y
             ||
             ||
             ||
             z
```

where x and y are nodes in the graph on level 2 and z is a node in the graph on level 3. we essentially make edges parallel to each other so we can connect x to y where ever z most closely projects onto the shape between x and y. at any rate i dont know why you would ever see a transit node on level 2 in heirarchy builder. that shouldnt happen. the only thing you should see there are nodes that should be there but with transit connection edges leaving them and going to level 3 tiles with transit nodes in them

cc @dnesbitt61 @gknisely who implemented more of this directly than i did

**Comment by mesozoic-drones** [comment]
Kevin, thank you for explanation about the transit connections!

>  i would at least try that just to see what happens. 

I tried this fast fix:
```
        if (directededge->bss_connection() || (directededge->use() == Use::kTransitConnection ||
                                               directededge->use() == Use::kEgressConnection ||
                                               directededge->use() == Use::kPlatformConnection)) {
          // Despite the road class, Bike Share Stations' connections are always at local level
          levels[2] = true;
        }
```

And it leads to the same `std::logic_error: current_level was never set`. 
if I also add upward transitions also for level 3:

 ```} else if (current_level >= 2) {```
https://github.com/valhalla/valhalla/blob/master/src/mjolnir/hierarchybuilder.cc#L343
 
then valhalla_build_tiles crashes on the next stages - on building shortcuts or if shortcuts are disabled on validating tiles.

```
terminating with uncaught exception of type std::runtime_error: /Users/o.khlopkova/rose/valhalla/baldr/graphtile.h:174 GraphTile NodeInfo index out of bounds: 737510,3,639 nodecount= 21
```

**Comment by pranavpandey1998official** [comment]
@mesozoic-drones can you give me entire log of valhalla_build_tiles leading upto the crash

**Comment by mesozoic-drones** [comment]
@pranavpandey1998official Here is full log of `$ ./valhalla_build_tiles -c /home/ail/valhalla.json /home/ail/tiles/san_francisco.osm.pbf`:
[crash.log](https://github.com/valhalla/valhalla/files/8176174/crash.log)

san_francisco.osm.pbf is extracted from [Northern California](http://download.geofabrik.de/north-america/us/california/norcal.html) with the help of osmium extract tool. The crash also reproduces on full file [norcal-latest.osm.pbf,](http://download.geofabrik.de/north-america/us/california/norcal-latest.osm.pbf) but it takes more time.

**Comment by pranavpandey1998official** [comment]
@mesozoic-drones , I got the bug I think, so overall your graph reader is also picking up transit tile , but it should only pick local tiles for hierarchy building and everything else hence you are getting error

this is where we are fethcing all local tiles
https://github.com/valhalla/valhalla/blob/72b8c6ca4be80bf7cd184157857e35f651a1785b/src/mjolnir/hierarchybuilder.cc#L418

can you log and confirm that the function is also returning transit tile
I am too lazy to setup everything, maybe we can debug this together as you already seems to have the setup, 
mail me pranavpandey1998official@gmail.com if you want to set up a meet

**Comment by pranavpandey1998official** [comment]
cool, now filter the transit tiles here and see if it works
https://github.com/valhalla/valhalla/blob/72b8c6ca4be80bf7cd184157857e35f651a1785b/src/mjolnir/hierarchybuilder.cc#L420

**Comment by mesozoic-drones** [comment]
I skipped all transit tiles in `CreateNodeAssociations()`:
```
    if (base_tile_id.level() == 3) {
      continue;
    }
```

After this change `valhalla_build_tiles` finished it's work without errors. It generated level 3 of tiles, all transit tiles are not empty.

But I can't build any transit route. For example if I send request with this json to /route
```json
{"locations":[{"lat":37.712757, "lon":-122.465103}, {"lat":37.796302, "lon":-122.394947}], "costing":"multimodal","date_time":{"type":0}}
```

valhalla returns error `{"error_code":440,"error":"Cannot reach destination - too far from a transit stop","status_code":400,"status":"Bad Request"}`.

I investigate this issue and for now I have a suspicion that the problem lies in tiles preparation, not in `valhalla_service`.

**Comment by mesozoic-drones** [comment]
I also checked transit with the help of /transit_available with 10 km radius:
```json
{"locations":[{"lat":37.712757, "lon":-122.465103, "radius":10000}]}
```

And `valhalla_service` returns istransit=false:
```json
[{"radius":10000,"input_lon":-122.465103,"istransit":false,"input_lat":37.712757}]
```

**Comment by pranavpandey1998official** [comment]
You should probably check the tradit and specify time too

**Comment by kevinkreiser** [comment]
ha that was going to be my other guess as well, that i probably changed the tileset function to return level 3 tiles as well.. apologies about that. lets PR that fix at least.

@pranavpandey1998official thats a good guess about the schedules. you could hack the dates though at import time just to check if it will work.

**Comment by mesozoic-drones** [comment]
I tried various requests with older dates. I set "date_time/value" to different dates in the end of 2020 and in the 2021 year. For example:

```json
{"locations":[{"lat":37.712757, "lon":-122.465103}, {"lat":37.796302, "lon":-122.394947}], "costing":"multimodal","date_time":{"type":1,"value":"2021-01-29T08:00"}}
```

For each request valhalla returned
```json
{"error_code":440,"error":"Cannot reach destination - too far from a transit stop","status_code":400,"status":"Bad Request"}
```

**Comment by mesozoic-drones** [comment]
As far as I know the /transit_available API doesn't depend on dates and schedules. Is it so?

**Comment by kevinkreiser** [comment]
True it doesn't not care about dates only that tiles exist near the location

**Comment by mesozoic-drones** [comment]
Ok. But /transit_avaliable doesn't detect any transit in tiles. So the problem isn't in old schedules.

**Comment by kevinkreiser** [comment]
That sounds like the case yep

**Comment by mesozoic-drones** [comment]
In valhalla_service in `loki_worker_t::transit_available(Api& request)` the `found` container always stays empty because `color` = 0:

https://github.com/valhalla/valhalla/blob/72b8c6ca4be80bf7cd184157857e35f651a1785b/src/loki/transit_available_action.cc#L39-L44

**Comment by kevinkreiser** [comment]
are there level 3 tiles in your tileset? if so then perhaps we modified the connectivity map in such a way as to have broken this (totally possible we badly need transit unit tests..)

**Comment by mesozoic-drones** [comment]
Yes, level 3 exists in the tileset. Transit tiles are not empty (200-400 KB).
